tools/jvm/src/main/scala/org/scalajs/core/tools/io/AtomicFileOutputStream.scala

```diff
@@ -0,0 +1,37 @@
+package org.scalajs.core.tools.io
+
+import java.io._
+import java.net.URI
+
+/** Provides best-effort atomic writing to a file
+ *
+ *  This FileOutputStream writes to a temporary file and moves it onto
+ *  the original file, once this FileOutputStream is closed. However,
+ *  it is only able to do so, if the underlying filesystem supports
+ *  it. If it doesn't it falls back to non-atomic moving or copying.
+ */
+private[io] class AtomicFileOutputStream private (
+    private val baseFile: File,
+    private val tmpFile: File
+) extends FileOutputStream(tmpFile) {
+
+  def this(baseFile: File) = {
+    this(baseFile, {
+      // Create a temporary file we actually write to
+      val tmpFile = File.createTempFile(
+          ".tmp-", baseFile.getName, baseFile.getParentFile)
+      tmpFile.deleteOnExit()
+      tmpFile
+    })
+  }
+
+  override def close(): Unit = {
+    super.close()
+
+    if (!tmpFile.renameTo(baseFile)) {
+      // Renaming failed. Fallback to copy
+      IO.copyTo(FileVirtualBinaryFile(tmpFile),
+          WritableFileVirtualBinaryFile(baseFile))
```

Comment:
Shouldn't we delete the temporary file after `copyTo`? I know it will be deleted on exit, but exit might be in a very long time. So not deleting it looks like a resource leak to me.